### PR TITLE
feat: add LSTM as optional forecasting algorithm (#133)

### PR DESCRIPTION
**Pull request by sufirehman**
## Summary

Adds a PyTorch LSTM model as an optional alternative to XGBoost for hourly electricity demand forecasting, addressing issue #133. Submitted as a draft because all items on the requested checklist are now complete, including the XGBoost-vs-LSTM benchmark on real DemandCast data.

## Checklist against @eantonini's requested items

- [x] Same public model interface as xgboost.py (train, predict, save, load, get_initialized_model)
- [x] Validated lstm_config.yaml following the existing configuration pattern (Pydantic ConfigModel, REQUIRED/OPTIONAL inline comments)
- [x] Unit tests consistent with the current testing approach (18 tests in tests/test_lstm.py covering fit/predict, entity-aware sequence construction, group boundary enforcement, save/load round-trip; 2 new tests in tests/test_ml.py; all passing)
- [x] Documentation in webpage/docs/ML.md: full LSTM Algorithm section covering motivation, architecture, config parameter table, dependencies, input/output format, and assumptions/limitations, with citation to the peer-reviewed methodology (Ul Rehman and Iqbal, 2025, doi:10.1186/s42162-025-00587-5)
- [x] Benchmark against XGBoost on real DemandCast data (DNK, AUT, PRT, 2021-2023) using identical train/validate/test logic and MAPE evaluation metric
- [x] LSTM kept as optional — XGBoost remains the default algorithm
- [x] Reasonable CPU-friendly defaults (n_timesteps=24, n_units=32, epochs=5) confirmed to train in under 15 minutes on CPU

## Benchmark results

Both algorithms trained and validated on Denmark, Austria, and Portugal (2021-2023) under identical conditions. Testing set MAPE:

| Entity | XGBoost | LSTM |
|--------|---------|------|
| DNK | 0.0931 | 0.1555 |
| AUT | 0.1085 | 0.1754 |
| PRT | 0.0869 | 0.1634 |
| Mean | 0.0962 | 0.1648 |

XGBoost outperforms LSTM on this benchmark. The most likely explanation is dataset scale: with only three entities and two years of hourly data, there are too few independent sequences for the LSTM to learn its recurrent weights without underfitting (training and testing MAPE are close for LSTM, ruling out overfitting). XGBoost benefits from the engineered temporal features that hand it context the LSTM must learn implicitly from data volume it does not yet have here. The full analysis is in the Benchmark Results section of ML.md.

XGBoost remains the recommended default. LSTM is included as an optional alternative for users who want to explore sequence models on larger datasets.

## Additional changes in this PR

Two small independent fixes included alongside the LSTM feature:

UTF-8 encoding fix: five places in utils/ (config.py x2, entities.py, ml.py, uploader.py) opened YAML config files without explicit encoding, falling back to the Windows locale default (cp1252). One config file contains a UTF-8 character (Kantō in tepco.yaml) that cp1252 cannot decode, causing a crash on Windows. Added encoding="utf-8" to all five. Full test suite still passes.

.gitignore: added .claude/ to prevent local Claude Code scaffolding from being accidentally committed.

## Dependencies

torch>=2.0.0 added to pyproject.toml with the PyTorch CPU wheel index. CPU-only installation is sufficient. No GPU

### CODE REVIEW

File: demandcast/cross_validate.py [hunk 1]
```diff
@@ -95,13 +96,15 @@ def _cross_validate(
...
     else:
         raise ValueError(f"Unsupported algorithm: {algorithm}")
 
     # Perform Leave-One-Group-Out cross-validation
     cv_results = cross_validate(
-        xgb_model,
+        model,
         prepared_dataset["features"],
         prepared_dataset["target"],
         groups=prepared_dataset["group"],
```
Comment (eantonini) [inline demandcast/cross_validate.py:123]: Heads up: `cross_validate()` from scikit-learn only ever uses `groups` to build the CV folds: it doesn't forward it into `model.fit()`, and that's true for any estimator, not just this one. It was never an issue before because this function only supported XGBoost, whose `fit()` doesn't use `groups`. Now that LSTM is plugged in here too, it needs `groups` during `fit()` to avoid mixing rows from different countries but it never receives it through this path. I checked this by patching `_make_sequences` and running this exact code: the LSTM always gets `groups=None` here, so cross-validation ends up treating the whole training fold as one continuous sequence. This needs a fix before merging, e.g. building the LOGO folds manually and calling `ml_models.lstm.train()`/`predict()` directly for the LSTM branch, bypassing scikit-learn's `cross_validate()` for this algorithm.

File: demandcast/ml_models/lstm.py [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+
+    Raises
+    ------
+    FileNotFoundError
+        If ``model_path`` does not exist.
+    """
+    if not os.path.exists(model_path):
+        raise FileNotFoundError(f"Model checkpoint not found: {model_path}")
+
+    checkpoint = torch.load(model_path, weights_only=False)
```
Comment (eantonini) [inline demandcast/ml_models/lstm.py]: `weights_only=False` allows arbitrary object deserialization when loading a checkpoint, which is a known security risk if a `.pt` file ever comes from somewhere untrusted. I tested saving/loading one of your checkpoints with `weights_only=True` and it worked without any changes. Could this be switched to `True` (or dropped, since that's now the default in recent PyTorch versions)?

File: demandcast/pyproject.toml [hunk 3]
```diff
@@ -40,10 +40,19 @@ dependencies = [
     "timezonefinder>=6.5.9",
     "tqdm>=4.67.1",
     "xarray>=2025.1.2",
+    "torch>=2.0.0",
```
Comment (eantonini) [inline demandcast/pyproject.toml]: Since LSTM is optional and not the default algorithm, would it make sense to move `torch` into an optional dependency group (e.g. `[project.optional-dependencies]`) rather than the main list? As it stands, every user, including anyone who only wants XGBoost, will now have to install PyTorch via `uv sync`. It also looks slightly inconsistent with `tests/test_lstm.py`, which uses `pytest.importorskip("torch", ...)` as if it expects torch might not be installed.

File: demandcast/ml_models/lstm.py [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+# Windows: CPython GC calls RemoveDllDirectory when the token from
+# add_dll_directory is collected; storing the token prevents removal.
+# torch must be imported before pandas — pandas side-effects corrupt
+# the Windows DLL loader state for torch's c10.dll initialisation.
+_dll_dir_tokens: list = []
+if sys.platform == "win32" and hasattr(os, "add_dll_directory"):
+    for _sp in sys.path:
+        _torch_lib = os.path.join(_sp, "torch", "lib")
+        if os.path.isdir(_torch_lib):
+            _dll_dir_tokens.append(os.add_dll_directory(_torch_lib))
```
Comment (eantonini) [inline demandcast/ml_models/lstm.py]: This Windows DLL workaround is duplicated almost line-for-line in `tests/conftest.py`. Could it be pulled into a single small shared helper (e.g. in a `utils` module) instead of copy-pasted? That way a future fix only needs to happen in one place.

File: demandcast/tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+if sys.platform == "win32" and importlib.util.find_spec("torch") is not None:
+    _dll_dir_tokens: list = []
+    if hasattr(os, "add_dll_directory"):
+        for _sp in sys.path:
+            _torch_lib = os.path.join(_sp, "torch", "lib")
+            if os.path.isdir(_torch_lib):
+                _dll_dir_tokens.append(os.add_dll_directory(_torch_lib))
+                break
+    import torch  # noqa: F401
```
Comment (eantonini) [inline demandcast/tests/conftest.py]: Same block as in `ml_models/lstm.py` (see comment there), worth factoring out into one shared function.

File: demandcast/ml_models/lstm.py [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+                i if groups[i] != groups[i - 1] else run_start[i - 1]
+            )
+
+        X_seq = np.zeros(
+            (n_samples, self.n_timesteps, n_features), dtype=X.dtype
+        )
+        for i in range(n_samples):
+            window_start = max(run_start[i], i - self.n_timesteps + 1)
+            n_valid = i - window_start + 1
+            X_seq[i, self.n_timesteps - n_valid :] = X[window_start : i + 1]
```
Comment (eantonini) [inline demandcast/ml_models/lstm.py:202]: Minor performance note: this loop builds each row's lookback window one at a time in plain Python. It's fine for the 3-country benchmark, but demandcast's full retrieval covers 40+ countries over multiple years of hourly data, so this could get slow at full scale compared to the vectorized version used in the no-groups branch above.

File: demandcast/benchmark_retrieve.ps1 [hunk 7]
```diff
@@ -0,0 +1,53 @@
+﻿# benchmark_retrieve.ps1 — fail-fast retrieval for DNK/AUT/PRT 2021-2023
+# Monitor: Get-Content -Wait logs\benchmark_orchestration.log
+
+Set-Location "C:\Users\Dell\Desktop\DemandCast\demandcast\demandcast"
```
Comment (eantonini) [inline demandcast/benchmark_retrieve.ps1]: This hardcodes your personal local path, probably not meant to be committed as-is? If these benchmark scripts are meant to help future contributors reproduce the benchmark, it'd be worth replacing this with a relative path or a note in the docs instead.

File: demandcast/benchmark_retrieve_resume.ps1 [hunk 8]
```diff
@@ -0,0 +1,46 @@
+# benchmark_retrieve_resume.ps1 — resume from step 5/6 only
+# Steps 1-4 already completed; this runs temperature (ERA5) + assemble.
+# Monitor: Get-Content -Wait logs\benchmark_orchestration.log
+
+Set-Location "C:\Users\Dell\Desktop\DemandCast\demandcast\demandcast"
```
Comment (eantonini) [inline demandcast/benchmark_retrieve_resume.ps1]: Same hardcoded personal path as in `benchmark_retrieve.ps1`, see comment there.

**Review by eantonini — CHANGES_REQUESTED** [review]
Thanks a lot for adding LSTM as a second algorithm. Overall, the interface mirrors the existing XGBoost module, the test suite is thorough (48 tests, including one specifically checking that sequences don't leak across country boundaries), and the write-up in `webpage/docs/ML.md` is honest about LSTM currently underperforming XGBoost on this dataset rather than overselling the result.

I tested the PR locally (installed dependencies, ran the new test suite, ran the linter) and dug into how the LSTM model behaves inside our actual pipeline scripts, not just in isolation. Everything I ran passed:
- All 48 new/changed tests pass.
- The exact ruff version and settings our pre-commit uses report no issues on any changed file.

However, I found one real bug that needs fixing before this can be merged, plus a handful of smaller suggestions. Details in the inline comments below, but here's the summary:

**1. Plugging LSTM into the existing cross-validation call doesn't carry the entity-boundary protection through (blocking)**
The main design idea of the LSTM model is that it shouldn't build a prediction window using rows from two different countries. That works correctly when calling `train()`/`predict()` directly. The issue is in how it's wired into `cross_validate.py`: scikit-learn's `cross_validate()` only ever uses the `groups` argument to decide how to split the data into folds; it has never forwarded `groups` into the model's `fit()` method, for any algorithm. That was never a problem before, because `cross_validate.py` only supported XGBoost, and XGBoost's `fit()` doesn't need `groups` at all (it treats every row independently). This PR is the first time an algorithm that does need `groups` during `fit()` (to avoid mixing rows from different countries) gets dropped into that same `cross_validate()` call, without adapting the call for that dependency. I confirmed the effect by patching `_make_sequences` and running it through `cross_validate.py`'s actual code path: the LSTM always receives `groups=None` there, so during cross-validation it silently treats the whole training fold as one continuous sequence, crossing country boundaries. It doesn't crash or warn, so the cross-validation MAPE numbers it produces would look normal but be unreliable.

**2. Loading a trained model uses an unnecessarily risky option (security)**
`torch.load(model_path, weights_only=False)` in `load()` allows loading arbitrary Python objects, not just model weights: a known risk if a checkpoint file is ever untrusted. I tested it and `weights_only=True` works fine for the checkpoint format you're using, so I don't think `False` is actually needed here.

**3. `torch` is a required dependency for everyone, even if they never use LSTM**
Right now `torch>=2.0.0` was added directly to `pyproject.toml`'s main dependency list, so every user (including anyone who only ever wants XGBoost) will download/install PyTorch. Since LSTM is optional and off by default, and your own tests already skip gracefully if torch isn't installed (`pytest.importorskip`), it'd be more consistent to make torch an optional dependency instead.

**4. A couple of smaller things** (see inline comments): the Windows-specific workaround code is duplicated in two files instead of being written once, the two `.ps1` benchmark scripts have your personal local Windows path hardcoded in them, and there's a performance note on the sequence-building loop for large datasets.

**One more thing worth double-checking (not something I could verify from the PR alone):** the entity-boundary logic assumes each country's rows always appear together as one contiguous block in the dataset. I looked at `assemble.py` (outside this PR) and didn't see an explicit sort step before the final data is merged and written out. It's possible this is already guaranteed elsewhere, but it'd be worth confirming that countries never get interleaved in the assembled dataset, since the LSTM code would silently misbehave (not crash) if they did.

Happy to help work through a fix for the cross-validation issue if useful. One option would be having `cross_validate.py` build the LOGO folds itself and call `train()`/`predict()` directly for the LSTM branch instead of going through scikit-learn's `cross_validate()`.

Thanks again for the contribution!

**Comment by sufirehman** [comment]
Hi @eantonini, all six points addressed and pushed:

1. Cross-validation groups bug (blocking): split _cross_validate into _cross_validate_xgboost (unchanged sklearn path) and _cross_validate_lstm, which builds LOGO folds manually with LeaveOneGroupOut().split() and calls ml_models.lstm.train()/predict() directly per fold, threading groups explicitly into fit() for each fold. A _GroupAwareLSTM adapter lets the existing scoring_metric config work generically. Added a regression test confirming each fold's training data excludes the held-out entity. Note: _cross_validate_lstm trains one model per fold sequentially with no parallelism, consistent with how cross_validate.py currently works.

2. Security: torch.load(..., weights_only=True). Verified via existing load round-trip test.

3. Optional dependency: torch moved to [project.optional-dependencies] lstm in pyproject.toml. uv.lock updated.

4. DLL deduplication: new utils/torch_windows.py with enable_torch_dll_directory(), imported from both ml_models/lstm.py and tests/conftest.py. Five new tests, 100% coverage on the new module.

5. Hardcoded paths: both .ps1 scripts now use $PSScriptRoot.

6. Performance note: comment added above the per-row loop in _make_sequences().

131 tests pass, ruff and mypy clean on all touched files. Two pre-existing failures in test_ml.py (Windows path separator assertions) are unrelated and pass in CI.

On the entity-contiguity question you raised: happy to add an explicit sort step in assemble.py if you'd prefer that over a documented assumption. Let me know which way you want to go.
